### PR TITLE
Added the ability to limit the number of items to parse from the XML tree

### PR DESCRIPTION
**Pull request by ccarnino**
I have added the Parser option to limit the number of items to parse in the XML tree. So when the items tree is too big, we can stop before. By default it's disabled.

I am using this library in my product.
I use the content-grabber option, so the time it takes to get all the items in the tree is considerable.

I was parsing this feed `http://www.sketchappsources.com/rss.xml` that has 1325 items in the tree. Well it  doesn't end parsing, because it takes way too long.
With the limit option I can limit the number of items parsed to something more reasonable.

### CODE REVIEW

File: lib/PicoFeed/Parser/Parser.php [hunk 1]
```diff
@@ -98,6 +98,14 @@
     private $grabber_ignore_urls = array();
 
     /**
+     * Limit the number of items to parse from the tree
+     *
+     * @access private
+     * @var bool
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: wrong type

File: lib/PicoFeed/Parser/Parser.php [hunk 2]
```diff
@@ -98,6 +98,14 @@
...
+     */
+    private $limit_number_items_to_parse = 0;
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: IMHO max_items is a more catchy name. I don't think that 0 is a good initial value. This way it wont be possible to disable item parsing at all.

File: lib/PicoFeed/Parser/Parser.php [hunk 3]
```diff
@@ -161,8 +169,14 @@ public function execute()
         $this->findFeedLogo($xml, $feed);
         $this->findFeedIcon($xml, $feed);
 
+        $indexItemTree = 0;
+
         foreach ($this->getItemsTree($xml) as $entry) {
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: What about replacing the foreach-loop with a for-loop, which runs till one of the following conditions is true:
- the current index is equal the number of items to parse
- last feed item is reached

File: lib/PicoFeed/Parser/Parser.php [hunk 4]
```diff
@@ -421,6 +436,19 @@ public function setGrabberIgnoreUrls(array $urls)
     }
 
     /**
+     * Set the limit of number of items to parse. Limit is active if bigger than 0.
+     *
+     * @access public
+     * @param  bool  $limit   Number
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: wrong datatype

File: lib/PicoFeed/Parser/Parser.php [hunk 5]
```diff
@@ -421,6 +436,19 @@ public function setGrabberIgnoreUrls(array $urls)
...
+     * @return \PicoFeed\Parser\Parser
+     */
+    public function setLimitNumberItemsToParse($limit)
+    {
+        $this->limit_number_items_to_parse = $limit;
+        return $this;
```
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: Why does this setter need to return the parser object?

Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: @fguillot Maybe I don't understand the purpose, but the return values of ```setHashAlgo()```, ```setTimezone()```... seam to me unnecessary as well, aren't they?

**Comment by mkresin** [comment]
I've the impression that you try to fix your issue at the wrong place. I mean, parsing the whole feed takes less than two seconds on my computer, which look acceptable to me.

I think a real fix would be to limit the number of items which the content grabber processes. This way it would be possible to grab the content of not yet processed articles later.

File: lib/PicoFeed/Parser/Parser.php (hunk 1 above)
Comment (ccarnino) [inline lib/PicoFeed/Parser/Parser.php]: You're right.

File: lib/PicoFeed/Parser/Parser.php (hunk 2 above)
Comment (ccarnino) [inline lib/PicoFeed/Parser/Parser.php]: I am not good at naming, so it's a good suggestion.

File: lib/PicoFeed/Parser/Parser.php (hunk 3 above)
Comment (ccarnino) [inline lib/PicoFeed/Parser/Parser.php]: I didn't want to impact the code too much, that's why I kept the foreach-loop.
Anyway I can do that.

File: lib/PicoFeed/Parser/Parser.php (hunk 4 above)
Comment (ccarnino) [inline lib/PicoFeed/Parser/Parser.php]: Again same typo as before, due to a fast copy-and-paste.

File: lib/PicoFeed/Parser/Parser.php (hunk 5 above)
Comment (ccarnino) [inline lib/PicoFeed/Parser/Parser.php]: I wanted to be consistent with the other methods.
Returning this allow you to chain all the method setter in your app. It creates a simpler API for the end user/dev.

**Comment by ccarnino** [comment]
I feel like limiting the number of items who get the content grabber processing is wrong.

If this parser return an array of items, you want to have all items processed the same way. You don't want to have some of them processed in a way, and some others in a different way.

Imagine how this would add complexity in using that array.

We agree that what make it slow is the content grabber on the items, but I still think that the limit should be on the number of items who will be processed, not on the ones who get content-grabber-processed.  IMHO.

File: lib/PicoFeed/Parser/Parser.php (hunk 5 above)
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: Thanks for the explaination. I never considered to use method chaining for setter. Smells like 
poorly readable code,

@fguillot We should add the return to either all setter (```disableContentFiltering()```, ```enableContentGrabber()``` and so on) or remove it from all setter. What should be done?

File: lib/PicoFeed/Parser/Parser.php (hunk 3 above)
Comment (mkresin) [inline lib/PicoFeed/Parser/Parser.php]: Well it's only a proposal. IMHO the code is easier to read with a for-loop. If you don't agree with that, please wait till someone else add a comment regarding this. It's not unlikely that I'm wrong with it.

**Comment by mkresin** [comment]
> I feel like limiting the number of items who get the content grabber processing is wrong.

Well, I'm wasn't exact with this. Limit the number of items for witch content get fetched from remote resources.

> If this parser return an array of items, you want to have all items processed the same way. 

My point is, that the items are not necessarily being processed the same way right now. Albeit it doesn't work completely as intended, you can at least:

- limit the fetching of remote content to case where a "rule file" for the remote site exists
- you can rewrite/replace the content of the feed items without fetching any remote content

It doesn't make sense to me to apply the max_items limit for the above examples. Or at least it doesn't make sense because fetching/parsing of remote website content takes to much time.

Same as before: If you don't agree with that, please defer any changes to your code till someone else add a comment regarding this. It's not unlikely that I'm wrong with it.

@Raydiation Would you have a look at this (including the outdated comments)?

**Comment by Raydiation** [comment]
The thing is that limiting the amount of fetched items could cause data loss if it's a bigger feed. IMHO the proper solution for this is to contact the website owner to support http caching and/or solve this in specific rule files

**Comment by mkresin** [comment]
Thanks for Feedback Bernhard!

I've the impression you missed the point. The problem aren't the 1300 items
of the feed. Those can be parsed without any problems.

The problem of the OP is that for each of the 1300 items the corresponding
website is grabbed when he enables the content grabber for the feed and
fetches the feed the first time. And grabbing 1300 websites takes a huge
amount of time...

I don't see how http caching or a specific rule file could solve the
problem.

**Comment by Raydiation** [comment]
Yes got that :)

The thing is that we could do that with rule files if they were loaded earlier.

**Comment by mkresin** [comment]
Got it! Indeed, an interesting approach. This would even allow a "per feed tuning".

**Comment by ccarnino** [comment]
I am not sure of following the conversation, I am not an expert of this project. It is something that I am using in my product and I love since it's simple enough.

I understand that I don't know the architecture of the library so I don't want to debate too much on what is the right solution, because at the end of the day, you know better than me.

I would just ask: is this a problem you can consider to handle?
My app is like blog reader, so the user add his feed, so I need to be able to limit the number of items are fetched from the feed a-priori. 

Thanks for the help anyway.

**Pull request closed without merging**